core/market/decentralized/src/protocol/discovery.rs

```diff
@@ -34,21 +46,182 @@ pub enum DiscoveryInitError {
 
 /// Responsible for communication with markets on other nodes
 /// during discovery phase.
-#[async_trait]
-pub trait Discovery: Send + Sync {
-    async fn bind_gsb(&self, prefix: String) -> Result<(), DiscoveryInitError>;
+#[derive(Clone)]
+pub struct Discovery {
+    inner: Arc<DiscoveryImpl>,
+}
+
+pub struct DiscoveryImpl {
+    offer_received: HandlerSlot<OfferReceived>,
+    retrieve_offers: HandlerSlot<RetrieveOffers>,
+}
+
+impl Discovery {
+    fn new(mut builder: DiscoveryBuilder) -> Result<Discovery, DiscoveryInitError> {
+        let offer_received = builder.offer_received_handler()?;
+        let retrieve_offers = builder.retrieve_offers_handler()?;
+
+        let inner = Arc::new(DiscoveryImpl {
+            offer_received,
+            retrieve_offers,
+        });
+        Ok(Discovery { inner })
+    }
 
     /// Broadcasts offer to other nodes in network. Connected nodes will
     /// get call to function bound in DiscoveryBuilder::bind_offer_received.
-    async fn broadcast_offer(&self, offer: Offer) -> Result<(), DiscoveryError>;
-    async fn retrieve_offers(&self) -> Result<Vec<Offer>, DiscoveryError>;
+    pub async fn broadcast_offer(&self, offer: Offer) -> Result<(), DiscoveryError> {
+        broadcast_offer(self.inner.clone(), offer).await
+    }
+
```

Comment:
Ok, I was wrong.
With some effort I managed to handle this with single function. The problem was with binding handlers. Bound function must consume self, otherwise it exceeds it's lifetime. But since Discovery is copy-able, this will work fine.
Some functions will take `&self` other `self`, I hope it won't explode later ;)